cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/ProgramRunDispatcherTask.java

```diff
@@ -0,0 +1,152 @@
+import io.cdap.cdap.app.runtime.ProgramController;
+import io.cdap.cdap.app.runtime.ProgramOptions;
+import io.cdap.cdap.app.runtime.ProgramRunnerFactory;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.internal.app.ApplicationSpecificationAdapter;
+import io.cdap.cdap.internal.app.deploy.ConfiguratorFactory;
+import io.cdap.cdap.internal.app.deploy.InMemoryProgramRunDispatcher;
+import io.cdap.cdap.internal.app.deploy.RemoteProgramRunDispatcherResponse;
+import io.cdap.cdap.internal.app.deploy.pipeline.ProgramRunInfo;
+import io.cdap.cdap.internal.app.runtime.artifact.ApplicationClassCodec;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactRepository;
+import io.cdap.cdap.internal.app.runtime.artifact.RequirementsCodec;
+import io.cdap.cdap.internal.app.runtime.codec.ArgumentsCodec;
+import io.cdap.cdap.internal.app.runtime.codec.ProgramOptionsCodec;
+import io.cdap.cdap.internal.io.SchemaTypeAdapter;
+import io.cdap.cdap.security.impersonation.Impersonator;
+import org.apache.twill.api.RunId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+
+/**
+ * Implementation of {@link RunnableTask} to execute Program-run operation in a system worker.
+ */
+public class ProgramRunDispatcherTask implements RunnableTask {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunDispatcherTask.class);
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(
+          new GsonBuilder())
+      .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+      .registerTypeAdapter(ApplicationClass.class, new ApplicationClassCodec())
+      .registerTypeAdapter(Requirements.class, new RequirementsCodec())
+      .registerTypeAdapter(RunId.class, new RunIds.RunIdCodec())
+      .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+      .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec())
+      .create();
+
+  private final Injector injector;
+
+  @Inject
+  ProgramRunDispatcherTask(Injector injector) {
+    this.injector = injector;
+  }
+
+  @Override
+  public void run(RunnableTaskContext context) throws Exception {
+    try {
+      ProgramRunInfo programRunInfo = GSON.fromJson(context.getParam(), ProgramRunInfo.class);
+      DispatchTaskRunner taskRunner = injector.getInstance(DispatchTaskRunner.class);
+      ProgramController programController = taskRunner.dispatch(programRunInfo);
+      RemoteProgramRunDispatcherResponse response = null;
+      if (Objects.nonNull(programController)) {
+        response = new RemoteProgramRunDispatcherResponse(true);
+      }
+      String result = GSON.toJson(response);
+      context.writeResult(result.getBytes(StandardCharsets.UTF_8));
+    } catch (Exception e) {
+      LOG.error("Exception: ", e);
```

Comment:
why do we want to catch and log?

Reply:
+1, doesn't this mean no response is returned?

Also, is this log viewable by the user somewhere or does it only end up in the task runner log?  Seems like it should at least write a RemoteProgramRunDispatcherResponse(false), but even better would be to include the exception in the response so that the caller can handle it appropriately.

Reply:
Fixed. If there is any issue, we throw an exception. The exception is then propagated all the way back to [AbstractProgramRuntimeService.java](https://github.com/cdapio/cdap/pull/14079/files#diff-67d24baf42393ca2d9e36afe6233ef1cc5bac4b49f8e1bf32d0acb92fbf532fb) and gets logged correctly in `programStateWriter.error(programRunId, e);`